namespace/bcl.adoc

```diff
@@ -0,0 +1,46 @@
+= Binary Compatibility Levels
+
+This is a classification of the levels of binary compatibility support for Java EE 8 classic applications in terms of the level of difficulty/complexity of the associated implementation and tools.
+
+== Level 0 - Jakarta EE 8 Only
+* No transformation required
+* Requires a Jakarta EE 8 container
+* No mandated support for backwards compatibility
+** Up to implementors to decide what level of support to provide
+
+== Level 1 - Jakarta EE 8 Translation to EE 9
+* Requires a direct translation of javax.* packages to jakarta.* analog
+* Requires transformation of META-INF/services to jakarta analogs
+* Requires transformation of properties files referencing javax based properties
+* Requires transformation of server specific properties that are classic javax based properties
+** This could be a specfic build or deployment step that processes an EE 8 app
```

Comment:
A build step would produce an EE9 binary that could live a long time... but there would be no source code and no migration path to EE9.1 or beyond for that binary.

A deployment step would have a startup time cost and the translation tools would need to be maintained for EE9.1 and beyond, which could become increasingly difficult.    If it is a deployment step, it should perhaps be deprecated from the start and produce a warning that the application should be updated.

Another option is a source generation tool to generate EE9 sourcecode from EE8 source code... but that is perhaps beyond our scope?

Reply:
Doing this at deployment time has the advantage of not requiring any tools to change
and paying the cost only once for the lifetime of the deployment.

Doing this at build time or run time has the advantage that it could be used by applications
outside of an application server, e.g., using "standalone" versions of the new jakarta.* APIs.

It's not clear to me why you think maintaining the translation tools will become increasingly
difficult.

If binary compatibility is defined by a spec, that spec might have limited lifetime, but I
would expect that vendors will need to maintain any translation tools for a very long time.

While source level tools might be outside our scope, we should definitely work with tool vendors to
create the refactoring tools that would allow this to be done at the source code level.